diffkemp/simpll/passes/FieldAccessFunctionGenerator.cpp

```diff
@@ -0,0 +1,191 @@
+        // (They also have to have the same debug location, i.e. correspond to
+        // a single field access in the C source code).
+        // A vector/stack is generated from each of them and the processStack
+        // function is called to further process the instructions.
+        for (auto &BB : Fun) {
+            for (auto &Inst : BB) {
+                if (isa<GetElementPtrInst>(Inst) && InstStack.empty()) {
+                    // Possible start of field access block.
+                    // Check if GEP has constant indices (this pass could
+                    // technically support all GEPs, but this is not implemented
+                    // yet)
+                    if (dyn_cast<GetElementPtrInst>(&Inst)->
+                            hasAllConstantIndices()) {
+                        DIL = Inst.getDebugLoc().get();
+                        InstStack.push_back(&Inst);
+                    }
+                } else if ((isa<GetElementPtrInst>(Inst) ||
+                           isa<CastInst>(Inst)) && !InstStack.empty()) {
+                   if (Inst.getDebugLoc().get() != DIL) {
+                       // Stop building stack.
+                       processStack(InstStack, Mod);
+                   } else {
+                       // Add instruction to stack.
+                       InstStack.push_back(&Inst);
+                   }
+                } else {
+                    // Wrong instruction type. Stop building stack.
+                    processStack(InstStack, Mod);
+                }
+            }
+            // Avoid processing instruction groups that contain instructions in
+            // two or more basic blocks.
+            processStack(InstStack, Mod);
+        }
+    }
+
+    return PreservedAnalyses();
+}
+
+/// First checks whether the instructions in the stack can be moved away from
+/// the original function without causing an error, then creates the abstraction
+/// function, moves the instructions into it, modify the input and output of the
+/// block to match the argument and returns value of the function and inserts
+/// the call to the newly create abstraction instead of the instruction block to
+/// the original function.
+void FieldAccessFunctionGenerator::processStack(
+        std::vector<Instruction *> &Stack, Module &Mod) {
+    if (Stack.empty())
+        // Do not attempt to process empty stacks, that would cause an error.
+        return;
+
+    // Check if all the instruction (besides the last one, of course) are used
+    // only by other instructions in the stack.
+    // If not, do not generate an abstraction, since this would break the code.
+    std::set<Value *> ValidUserSet;
+    for (Instruction *Inst : Stack)
+        ValidUserSet.insert(Inst);
+    bool allUsersValid = true;
+    for (Instruction *Inst : Stack) {
+        if (Inst == Stack.back())
+            continue;
+        for (User *U : Inst->users()) {
+            if (ValidUserSet.find(U) == ValidUserSet.end()) {
```

Comment:
Search in `Stack` (after making it constant) and directly return if user is not there.